File: lennard-jones/cpp/lj2d.cpp
// Input format:
// ./lj2d L N MCS T [U_minstep] [U_jump] [start_dist] [delta]

#include <iostream>
#include <iomanip>
#include <assert.h>
#include <random>
#include <math.h>
#include <string>
#include <time.h>
#include <forward_list>
using namespace std;
typedef unsigned int uint;

// GLOBAL CONSTANTS (to be read from parameters) ___
double L;
double delta;
double T;
uint N;
uint MCS;
double start_dist;

// GLOBAL ARRAYS ___________________________________
double* x;
double* y;
double* U;

// RANDOM NUMBER GENERATOR
// (returns random double x = [0.0, 1.0)) __________

uint seed = time(nullptr);
default_random_engine eng(seed);
uniform_real_distribution<double> dist(0., 1.);

double rndReal()
{
    return dist(eng);
}

// PERIODIC BOUNDARY CONDITIONS ____________________
double pbc(double q)
{
   double pbc = fmod(q,L);
   if(pbc < 0) pbc += L;
   return pbc; 
}

// CALCULATION OF ENERGY OF AN ATOM
// (after being moved by [dx,dy]) __________________
double energy(uint i, double dx = 0.0, double dy = 0.0)
{
    double Ui = 0;
    for(int j = 0; j < N; ++j)
    {
        if(i != j)
        {
            double dxij = pbc(x[i] + dx) - x[j];
            if(dxij > L/2) dxij = L - dxij;
            if(dxij < -L/2) dxij = L + dxij;

            double dyij = pbc(y[i] + dy) - y[j];
            if(dyij > L/2) dyij = L - dyij;
            if(dyij < -L/2) dyij = L + dyij;
            
            double r2ij = pow(dxij,2) + pow(dyij,2);
           
            Ui += 4.0*(pow(r2ij,-6) - pow(r2ij,-3));
        }
    }
    return Ui;
}

// CALCULATION OF VARIANCE _________________________
double variance(forward_list<double>* list)
{
    double sum = 0;
    uint count = 0;
    for(auto j = list->begin(); j != list->end(); ++j)
    {
        sum+=*j;
        ++count;
    }
    double avg = sum/count;

    double var = 0;
    for (auto j = list->begin(); j!= list->end(); ++j)
    {
        var += pow(*j-avg,2);
    }
    var /= count;

    return var;
    
}

// SINGLE MONTE CARLO STEP _________________________
void mcStep()
{
    for(uint i = 0; i < N; i++)
    {
        double dx = delta*(rndReal()-0.5);
        double dy = delta*(rndReal()-0.5);
        double u_new = energy(i,dx,dy);
        double dU = u_new - U[i];
        if(dU < 0  ||  rndReal() < exp(-dU/T))
        {
            U[i] = u_new;
            x[i] = pbc(x[i]+dx);
            y[i] = pbc(y[i]+dy);
        }
    }
    return;
}

int main(int argc, char** argv)
{
    assert(argc > 4);
    // READ INPUT PARAMETERS _______________________
    L = stod(argv[1]);
    N = stoi(argv[2]);
    MCS = stoi(argv[3]);
    T = stod(argv[4]);
    uint U_minstep =(argc > 5) ? stoi(argv[5]) : 0;
    uint U_jump = (argc > 6) ?  stoi(argv[6]) : 1;
    start_dist = (argc > 7) ? stod(argv[7]) : 0.3;
    delta = (argc > 8) ? stod(argv[8]) : 0.1;

    
    assert(L>0);
    assert(N>0);
    assert(MCS>=0);
    assert(T>0);

    double ro = (double)N / L / L;

    cout <<"# L = "<< L << "  N = " << N << "  MCS =" << MCS 
    << "  ro* = " << ro << "  T* = "<< T << endl
    << "# Seed = " << seed << endl;

    // GENERATE START CONFIGURATION ________________
    x = new double[N];
    y = new double[N];
    U = new double[N];
    forward_list<double> U_step;
    
    uint rows = sqrt(N);
    uint i = 0;
    for (uint j = 0; j < rows; j++)
    {
        for (uint k = 0; k < N/rows +1; k++)
        {
            if (i < N)
            {
                x[i] = start_dist*j;
                y[i] = start_dist*k;
                i++;
            }
            
        }
        
    }
    
    for (uint i = 0; i < N; i++)
    {
        U[i] = energy(i);
    }
    
    // RUN THE SIMULATION __________________________
    for(uint step=0; step<MCS; step++)
    {
        mcStep();

        if (step > U_minstep)
        {

            if ((step-U_minstep) % U_jump == 0)
            {
                double U_total = 0;
                for(uint i=0;i<N;i++) U_total += U[i];
                U_step.push_front(U_total);
            }
            
        }
    }
    double c = 1/(N*T*T) * variance(&U_step);
    // PRINT RESULTS _______________________________
    
    cout << "# x\t\t  y\t\t  U" << endl;
    cout << "# c = " << c <<endl;
    for (uint i = 0; i < N; i++)
    {
        cout << fixed << x[i]<< "\t"  << y[i] << "\t" << scientific << U[i] << endl;
    }

    
    delete x;
    delete y;
    delete U;
    return 0;
}

File: lennard-jones/python/lj2d.py
import numpy as np
import random as rnd
import math
# WEJŚCIE ________________________________________________________________

L = float(input())
N = int(input())
MCS = int(input())
T = float(input())
delta = float(input())
start_distance = float(input())

assert L > 0
assert N > 0
assert MCS > 0
assert T > 0
ro = N/L**2
assert ro <= 1.0

print("# L = {}   N = {}   ro* = {}   T = {}   MCS = {}   delta = {}".format(L,N,ro,T,MCS,delta))


# PRZYGOTOWANIE SYMULACJI ____________________________________________________

R = np.zeros((N,2))  # tablica złożona z wektorów [x,y] położeń atomów

rows = math.floor(N**0.5)  #rozmieszczenie atomów
i = 0
for ix in range(rows):
    for iy in range(N//rows + 1):
        if i < N:
            R[i] = [ix*start_distance,iy*start_distance]
            i += 1


def energy(i,dr):    # funkcja zwraca energię atomu o indeksie i, po przesunięciu o wektor dr
    ri = (R[i] + dr) % L
    Rj = np.delete(R,i,0)                    # kopia tablicy R bez elementu ri
    r_ij = np.linalg.norm(Rj - ri, axis = 1) # tablica odległości [r_i1, r_i2, r_i3, ..., r_iN]  #BUG - uwzględnić PBC!
    u = np.sum(4.0*(r_ij**(-12)-r_ij**(-6))) # suma po atomach rj (j != i)
    return u

U = [energy(i,0) for i in range(N)]

# SYMULACJA ___________________________________________________________________
for step in range(MCS):
    
    # KROK MONTE CARLO ________________________________________________________
    
    for i in range(N):
        dr = [delta*(rnd.random()-0.5), delta*(rnd.random()-0.5)]
        u_new = energy(i,dr)
        dU = u_new - U[i]

        # METROPOLIS __________________________________________________________
        if dU < 0:
            U[i] = u_new
            R[i] = (R[i] + dr) % L
        
        elif rnd.random() < math.exp(-dU/T):
            U[i] = u_new
            R[i] = (R[i] + dr) % L

for ri in R:
     print("{}\t{}".format(ri[0],ri[1]))